File: SearchEngine/Trie.js
class TrieNode {
  constructor() {
    this.children = Object.create(null);
    this.isEnd = false;
    this.words = new Set();  // limit suggestions by storing docs
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word, docId) {
    let node = this.root;

    for (const char of word) {
      if (!node.children[char]) {
        node.children[char] = new TrieNode();
      }
      node = node.children[char];
      node.words.add(docId);
    }

    node.isEnd = true;
  }

  // returns top N ranked suggestions
  suggest(prefix, limit = 10) {
    let node = this.root;

    for (const char of prefix.toLowerCase()) {
      if (!node.children[char]) return [];
      node = node.children[char];
    }

    return Array.from(node.words).slice(0, limit);
  }
}

module.exports = { Trie };

File: Go/SearchEngine-go/trie.go
package main

type TrieNode struct {
	Children map[rune]*TrieNode
	IsEnd    bool
	Words    map[string]bool
}

type Trie struct {
	Root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{Root: &TrieNode{Children: make(map[rune]*TrieNode), Words: make(map[string]bool)}}
}

func (t *Trie) Insert(word, docID string) {
	node := t.Root
	for _, char := range word {
		if node.Children[char] == nil {
			node.Children[char] = &TrieNode{Children: make(map[rune]*TrieNode), Words: make(map[string]bool)}
		}
		node = node.Children[char]
		node.Words[docID] = true
	}
	node.IsEnd = true
}

func (t *Trie) Suggest(prefix string, limit int) []string {
	node := t.Root
	for _, char := range prefix {
		if node.Children[char] == nil {
			return nil
		}
		node = node.Children[char]
	}
	var docs []string
	for docID := range node.Words {
		docs = append(docs, docID)
		if len(docs) >= limit {
			break
		}
	}
	return docs
}
